### ISSUE TITLE
Add command to disable the language server

### ISSUE THREAD

**Issue by Alizter** [ejgallego/coq-lsp#209]
It would be good if we could deactivate the entire extension using a command. Sometimes you might be editing a .v file without wishing to start coq-lsp, in these cases deactivating would be useful.

**Comment by ejgallego** [comment]
I'm not sure this is possible, unless you want actually a command to stop the language server only? If so, that'd be easier.

**Comment by Alizter** [comment]
Yes, I suppose that would do. At the moment, I end up having to fight lsp when I edit a .v file. Updating the cursor reopens the goals window after closing it, etc.

**Comment by ejgallego** [comment]
Ouch that would amount to a full disable of the extension; do you know that you can disable / enable extensions only for particular workspaces?

We could have such a command indeed, not sure how clean the implementation can be.

**Comment by ejgallego** [comment]
Maybe you are more bothered by the auto-follow; how would a command to "disable/enable auto info" would work for you?

**Comment by Alizter** [comment]
@ejgallego I've seen other extensions allow for a togglable state. You just run a command that turns off the entire extension, and then run the same command to turn it back on. For example, https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker allows you to toggle the spell checking extension. You don't need to mess around with workspaces AFAIK.

**Comment by ejgallego** [comment]
Sounds good to me, should we put the command in the status bar?

Indeed I see the feature better now, but we need a bit of code refactoring to make the disable be clean, #211 is a step for that.

We can already prototype the status bar button and command tho.

**Comment by Alizter** [comment]
rust-analyzer which is the lsp server/extension for rust has a feature like this. A button appears on the status bar which when clicked will stop the running of cargo (the build system) in the background. I think it would be a huge usability boost to have something like this for coq-lsp. If you are making small refactoring edits to 10s of .v files, you really feel LSP struggling.

Here restarting LSP is not an option since the process may not receive the termination signal in time, causing your computer to deadlock. But this is another issue.

**Comment by ejgallego** [comment]
> Here restarting LSP is not an option since the process may not receive the termination signal in time, causing your computer to deadlock.

How so? Stopping the extension should be fine, and indeed it is the only way to implement what you want?

Tho actually it seems to me that your problem is better solved by Pull Diagnostics.

**Comment by ejgallego** [comment]
But yes, it'd be nice to have a bit more control, the status bar idea is good like Rust (nice finding) however ideally we would call `client.stop()`.

What does Rust client do?

**Comment by Alizter** [comment]
>What does Rust client do?

I'm not sure about the details, however there are situations in Coq where we might never return from a process. i.e. having a typeclass search blow up. In this case, we need better interrupts in Coq so that the lsp process is not left doing nonsense work.

For rust, this is not a problem since I have never seen the checking diverge. There used to be cases with OCaml/merlin where checking a `rec` would diverge and I observed similar issues. I believe that was fixed by interrupting better (probably in merlin).

**Comment by ejgallego** [comment]
> I'm not sure about the details, however there are situations in Coq where we might never return from a process. i.e. having a typeclass search blow up. In this case, we need better interrupts in Coq so that the lsp process is not left doing nonsense work.

Restarting the extension solves this problem, doesn't it?

Yes, disabling the extension checking is a valid and desirable feature request, but indeed what I think is that the out of the box experience should be such that users never have to resort to it.

**Comment by Alizter** [comment]
See how rust analyzer does it: https://github.com/rust-lang/rust-analyzer/blob/368e0bb32f1178cf162c2ce5f7e10b7ae211eb26/editors/code/src/main.ts#L117

**Comment by ejgallego** [comment]
After #377 , all that should be missing for this issue to be solved is to add the status button.
